File: src/core/piece_table.zig
const std = @import("std");
pub const PieceSource = enum { Original, Add };

pub const Piece = struct {
    source: PieceSource,
    start: usize,
    length: usize,
};

pub const PieceTable = struct {
    original: []const u8,
    add: []u8,
    pieces: std.ArrayList(Piece),

    pub fn new(allocator: *std.mem.Allocator, original: []const u8) !PieceTable {
        var pieces = std.ArrayList(Piece).init(allocator.*);
        try pieces.append(Piece{ .source = .Original, .start = 0, .length = original.len });
        return PieceTable{ .original = original, .add = &[_]u8{}, .pieces = pieces };
    }

    pub fn insert(self: *PieceTable, pos: usize, text: []const u8) !void {
        const add_pos = self.add.len;
        self.add = try std.fmt.allocPrint(self.pieces.allocator, "{s}{s}", .{ self.add, text });
        var piece_idx = self.find_piece(pos);
        const piece = self.pieces.items[piece_idx];
        //incase of split
        if (piece.start + piece.length > pos) {
            const split_offset = pos - piece.start;
            const new_piece = Piece{ .source = piece.source, .start = piece.start + split_offset, .length = piece.length - split_offset };
            self.pieces.items[piece_idx].length = split_offset;
            try self.pieces.insert(piece_idx + 1, new_piece);
            piece_idx += 1;
        } else if (self.pieces.items.len <= 1) {
            //handle edge cases when less than 2 elements present in piecetable
            const new_piece = Piece{ .source = .Add, .start = add_pos, .length = text.len };
            try self.pieces.append(new_piece);
        } else {
            //insert a new piece for added text
            const new_piece = Piece{ .source = .Add, .start = add_pos, .length = text.len };
            try self.pieces.insert(piece_idx, new_piece);
        }
    }

    pub fn destroy(self: *PieceTable) void {
        self.pieces.deinit();
    }

    pub fn find_piece(self: *PieceTable, pos: usize) usize {
        var offset: usize = 0;
        for (self.pieces.items, 0..) |piece, i| {
            if (offset + piece.length > pos) {
                return i;
            }
            offset += piece.length;
        }
        return self.pieces.items.len - 1;
    }

    pub fn get_text(self: *PieceTable, allocator: *std.mem.Allocator) ![]u8 {
        var builder: []u8 = "";

        for (self.pieces.items) |piece| {
            const source = switch (piece.source) {
                .Original => self.original,
                .Add => self.add,
            };
            //try builder.appendSlice(source[piece.start .. piece.start + piece.length]);
            builder = try std.fmt.allocPrint(allocator.*, "{s}{s}", .{ builder, source[piece.start .. piece.start + piece.length] });
        }

        return builder;
    }
};

File: src/radiance/core/piece_table.rs
#[derive(Debug, Clone)]
pub enum PieceSource {
    Original,
    Add,
}
#[derive(Debug, Clone)]
pub struct Piece {
    start: usize,
    length: usize,
    source: PieceSource,
}
#[derive(Debug, Clone)]
pub struct PieceTable {
    pub original: String,
    pub add: String,
    pub pieces: Vec<Piece>,
}

impl PieceTable {
    pub fn new(original: String) -> PieceTable {
        let length = original.len();
        PieceTable {
            original,
            add: String::new(),
            pieces: vec![Piece {
                start: 0,
                length,
                source: PieceSource::Original,
            }],
        }
    }

    pub fn insert(&mut self, pos: usize, text: &str) {
        let add_pos = self.add.len();
        self.add += text;

        let piece_idx = self.find_piece(pos);
        let piece = self.pieces.get(piece_idx).unwrap();

        self.pieces.push(Piece {
            source: PieceSource::Add,
            start: add_pos,
            length: text.len(),
        });
    }

    pub fn pop(&mut self, pos: usize, length: usize) {
        let mut new_pieces = Vec::new();
        let mut current_pos = 0;

        for piece in self.pieces.clone() {
            if current_pos + piece.length <= pos {
                new_pieces.push(piece.clone());
            } else if current_pos >= pos + length {
                new_pieces.push(piece.clone());
            }
            current_pos += piece.length;
        }
        self.pieces = new_pieces;
    }

    fn find_piece(&self, pos: usize) -> usize {
        let mut offset = 0;
        for (idx, piece) in self.pieces.iter().enumerate() {
            if offset + piece.length > pos {
                return idx;
            }
            offset += piece.length;
        }
        self.pieces.len() - 1
    }

    pub fn get_text(&self) -> String {
        let mut result = String::new();

        for piece in self.pieces.iter() {
            let source = match piece.source {
                PieceSource::Original => &self.original,
                PieceSource::Add => &self.add,
            };
            result += &source[piece.start..(piece.start + piece.length)];
        }
        result
    }
}
